README.md
# hpyculator

 >high extensibility calculator base on python

![图标](ico.ico "随手画")

![GitHub](https://img.shields.io/github/license/HowieHz/hpyculator)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/hpyculator)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/48247dc302b44b1ebe1ca6635a7a0588)](https://www.codacy.com/gh/HowieHz/hpyculator/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=HowieHz/hpyculator&amp;utm_campaign=Badge_Grade)
![GitHub all releases](https://img.shields.io/github/downloads/HowieHz/hpyculator/total)
![GitHub repo size](https://img.shields.io/github/repo-size/HowieHz/hpyculator)

**这是一个基于python的高拓展性计算器**

hpyculator（下文简称hpyc），是我怀揣着 “一个一个添加算法太累了，要是能像加载Minecraft的mod一样加载算法就好了” 这样的想法

在[各类数组程序V1.2](https://www.bilibili.com/video/BV18p4y1h7bQ)的基础上开发而来
  
>本项目基于python3.10构建

### 使用场景

1.多功能工具箱

2.写了一个小工具，这时候只需要添加一个包含一些数据的字典和一个函数，就可以把它变成一个hpyculator插件来使用

## 开始

- **用户手册**: [UserManual](docs/source/user_manual.md)
- 开发文档: [开发插件](https://hpyculator.readthedocs.io/zh_CN/latest/quick_start.html#id3)或[参与到此项目](https://hpyculator.readthedocs.io/zh_CN/latest/quick_start.html#id9)
- 更新日志: [release](https://github.com/HowieHz/hpyculator/releases) 或 [changelog](https://hpyculator.readthedocs.io/zh_CN/latest/changelog.html)

[//]: # (- 开发计划: [milestones]&#40;https://github.com/HowieHz/hpyculator/milestones&#41;)

## 插件

[这里](https://github.com/HowieHz/PluginCatalogue) 是一个 hypc 的插件收集仓库（建设中）

可以先提交到本仓库，作为内置插件发行

## 文档

想要知道有关hpyculator的更多信息？ 去看[文档](https://hpyculator.readthedocs.io/)吧

文档中文版由HowieHz编写

## 其他

遇到问题或者有对项目发展的建议欢迎提issue

[项目计划](project_plan.md)

## 鸣谢

[感谢列表](https://hpyculator.readthedocs.io/zh_CN/latest/credits.html)

### 感谢以下项目

#### Qt实现无边框窗口

   [![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=zhiyiYo&repo=PyQt-Frameless-Window)](https://github.com/zhiyiYo/PyQt-Frameless-Window)

### 特别鸣谢

感谢 [JetBrains](https://www.jetbrains.com/?from=hpyculator) 为开源项目提供免费的 [IntelliJ PyCharm](https://www.jetbrains.com/pycharm/?from=hpyculator) 等 IDE 的授权 


docs/changelog.md
更新日志
====================================

    Added － 新增加的功能／接口
    Changed － 功能／接口变更
    Deprecated － 不建议使用的功能／接口，将来会删除
    Removed － 已删除的功能／接口
    Fixed － 解决的问题
    Others － 性能优化和安全性增强等改进

v1.7.2(草稿)
    None 

v1.7.1

    Added
        文档修改，补充
        添加hpyc_cli v1.0.0
    Changed
        修改一些插件的处理算法
    Fixed
        hpyc_pyside_ui达到输出上限不能正常停止消息监听线程的问题
    Others
        对模块名称进行了区分，防止意外的导入操作

v1.7.0
    
    Added
        分离核心为hpyc_core
        分离基于pyside6的ui为hpyc_pyside_ui
        完成文档关于API的说明
        hpyc_pyside_ui新增打表模式，每次输出后添加一个","
        新增common_factor_hz，pluperfect_digital_invariant_hz插件
    Changed
        多处文档更新
        插件更新
    Removed
        移除了output_without_line_break
        移除了hpyc.RETURN_ITERABLE_OUTPUT_IN_ONE_LINE
    Others
        代码性能优化（尽可能多的使用元组）
        结构优化，并补充注释，类型表示
        引入mypy进行静态类型检查

v1.6.3

    Added
        添加了output_without_line_break函数
    Fixed
        hpyc.RETURN_ITERABLE_OUTPUT_IN_ONE_LINE无法正常工作的错误
        test9_hz不正常工作的错误（表现为输出错误）
        output文档描述错误
        插件可能导致内存占用极大的特性（修复了文件缓冲不可控的问题）
        修复了插件缓冲和实际缓冲不符的问题
    Removed
        hpyc.RETURN_LIST
        hpyc.RETURN_LIST_OUTPUT_IN_ONE_LINE
    Others
        代码性能优化，结构优化，并补充注释

v1.6.2

    Added
        在搜索tag的时候会显示可用tag
    Changed
        常量名修改:
            hpyc.RETURN_LIST -> hpyc.RETURN_ITERABLE
            hpyc.RETURN_LIST_OUTPUT_IN_ONE_LINE -> hpyc.RETURN_ITERABLE_OUTPUT_IN_ONE_LINE
    Fixed
        翻译不正确显示的问题
        退出残留问题
    Deprecated
        hpyc.RETURN_LIST
        hpyc.RETURN_LIST_OUTPUT_IN_ONE_LINE
    Others
        文字归档到一个文件，易于翻译和管理
        迁移到3.10
        修改写法(if elif -> match case)
        优化嵌套结构

20220611
v1.6.1

    Added
        添加多种设置文件格式支持（json，toml，yaml）
    Changed
        默认设置文件格式从toml更改为json
    Removed
        移除软件内置的更新日志
        移除输入框输入update_log显示更新日志
    Others
        修改文档描述
        修改计算后输出时间统计格式和使用函数

20220529
v1.6.0-fix.1

    Fixed
        控件名错误的bug 
    Deprecated
        软件内置的更新日志

20220528
v1.6.0

    Added
        i18n -> en语言支持
        tag系统，可以使用tag进行搜索了
        新增插件若干
        可切换背景图片
        插件元数据tag项
        可切换是否自动换行
    Changed
        插件元数据option_name -> option
        插件元数据author现支持多作者
        文档示例更新
    Removed
        移除无用插件
        插件元数据use_quantifier项
    Fixed
        修复插件无法正常运行的bug
    Others
        代码风格修订

v1.5.0

    Changed
        重绘ui

20220430-20220505
1.4.0

    Changed
        修改了api和插件事件
    Removed
        移除test模式
    Others
        分离插件读取部分
        性能优化（干掉exec，解决命名空间污染）
        规范文档，规范版本，编程风格修订
        菜单栏字体大小调整
        调整搜索框大小
        文档更新(函数说明，示例调整，api，插件示例）

20220429
v1.3.3

    Others
        debug

20220429
v1.3.2

    Added
        添加了设置窗口
        修改保存位置的选项回归
    Fixed
        修复搜索框不可用的问题
    Others
        完成从wxpython移植到pyside6

20220416
v1.3.1

    Fixed
        debug

20220414
v1.3.0

    Others
        部分移植到pyside6上

20220412
v1.2.7

    Others
        程序优化
        主程序优化
        内置插件优化

20220411
V1.2.6

    Added
        增加了输出上限的选项
    Fixed
        修复了内置插件不能正常工作的问题
        修复遗留问题
    Others
        优化了输出优化的模式

20220409
V1.2.5

    修复了遗留问题
    完善了输出优化模式
    细节修复
    程序优化

20220408
V1.2.4

    hpyculator模块上传了！现可通过pip下载
    重写了内置插件
    修改了gui,可自选保存位置

20220408
V1.2.3

    修复了多线程的问题
    修改了菜单栏

20220407
V1.2.2

    修复了内置插件无法正常工作的问题
    添加了更新检查按钮(打开浏览器跳转到https://github.com/HowieHz/hpyculator/releases)

20220407
V1.2.1

    修复了无插件打不开的bug
    修复了退出后还残留后台程序的bug

20220407
V1.2.0

    修改gui
    增加了搜索框
    添加了测试选项，勾选之后程序就会使用test模式

20220405
V1.1.0

    增加了对文件插件的支持
    增加了很多提示，防止插件卡死程序

V1.0.0
改名为
皓式可编程计算器hpyculator

    2022年1月19日添加了规范文档
    
    2022年1月18-19日，重构了程序（从流水线+函数群变成了面向对象的编程）
    
    2022年3月29再次捡起项目，
    
    2022年3月31写完，实现了规范文档，但是掉进了打包的坑，还好第二天爬出来了
    
    2022年4月1日第一个内测版
    
    2022年4月1日添加了多线程支持，添加了更多用户友好类提示，修复了若干个bug，添加了一种输出模式
    
    2022年4月2日修改了内置插件的性能，添加了新的输出模式，并提供了4个函数便于插件作者创作

    趣闻:
    2022年1月16日时想用PyQt（5/6）代替wxpython（哈哈wxpython不支持python3.10，其他版本又删掉了，被迫的）
    由于PyQt（5/6）-tool用pip不能正确安装和考虑到学习成本，所以干脆卸载py310，改回py39（白配置一遍^_^）

来自 各类数组计算程序
2021年6月27日
V1.2

    选择保存文件时不再输出
    平均数，众数，中位数，方差，标准差的计算可以输入负数和小数了
    为了节省体积，换了个图标

V1.1

    添加了平均数，众数，中位数，方差，标准差的计算

2020年11月14日
V1.0

    多了一个可以切换计算核心的选项,
    更加模块化，内置模块有杨辉三角计算和斐波那契数列计算
    (作者电脑i5-5代低压u，4g内存。无聊算114514行，算了半小时，原本要关了，看任务管理器还在运行，就去睡了一会，起来看到一个1338412kb的文件（约1.28G），哈哈哈，请勿模仿（除非很闲）)

来自杨辉三角计算程序v3.4.2

    a3.4.2 修复不能读取配置的问题，优化代码体积
    
    2020年11月8日a3.4.1 创建了两个文件夹专门用来储存设置和输出,扩大1.5倍了默认生成的窗口
    
    2020年11月8日a3.4.0 在输入栏输入update_log点击运算就可以看更新日志了!菜单栏它来了
    
    2020年11月7日a3.3.4 修改保存名的时间
    
    a3.3.3b 走向规范化
    
    2020年11月7日a3.3.3a 遗弃的分支，消除了保存时间
    
    2020年10月17日a3.3.2 删除多余代码优化体积
    
    2020年10月17日a3.3.1 生成在gui里的不再是一行了,砍掉了可以修改算法的选项,删除多余代码优化体积
    
    a3.3.0 被遗忘的版本号
    
    a3.2 改进了算法，增快了运行速度,增加了可以修改算法的选项
    
    2020年10月10日a3.1 将原来的运行并保存按钮修改为是否保存的选项
    
    a3.0 有GUI了!生成的文件不再是一行了
    
    a2.0 添加了保存文件到本目录，添加了计算计时器,防止因为错误信息造成的闪退,节省打包体积
    
    a1.0 最初的经典(20220505注，最初的大粪)


src/Doc.py
import Version

TODO = """
更新展望(咕咕咕):
#todo 8.可以分享脚本的平台（qq群也不错） 17.Github actions
#todo 10.上线网页版
#todo 11.可以读取文件作为输入
#todo 15.美化 1.背景图
#todo 16.复制到剪贴板"""
# todo 8.可以分享脚本的平台（qq群也不错） 17.Github actions
# todo 10.上线网页版
# todo 11.可以读取文件作为输入
# todo 15.美化 1.背景图
# todo 16.复制到剪贴板


"""
各种量的命名规范
hello_world 变量全部小写，使用下划线连接
helloWorld 函数(def)和方法使用小驼峰式命名法，首单词字母小写，后面单词字母大写
HelloWorld 类名(Class)、文件名(Xswl.txt)使用帕斯卡命名规则(大驼峰式命名法,每一个单词的首字母都采用大写字母)。
HELLO_WORLD 常量(NEVER_GIVE_UP)全部大写，使用下划线连接单词
numba jax感觉两个差不多都用不了
"""

START_SHOW = ("""皓式可编程计算器 hpyculator %s
HowieHz制作

首先选择你要计算的东西
然后在上方输入框输入你要计算的项数（行数）
之后在左侧选择计算核心

必读:


！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！
0.建议保存到文件，这样不会内屏输出导致卡死!!--------重点！！！
！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！

1.“保存”会将结果保存到程序所在目录下 "皓式程序输出" 文件夹内
2.设置行(项)数较大请选择保存到文件，算的不久，但是导出很久，可以看看任务管理器，不同插件性能（读写速度和）不同，要看插件作者的水平

交流群694413711 - 此群安静的像一滩死水 -=作者中考完了，消息一般都会会

已在github开源 禁止商用






悄悄说:在输入栏输入update_log之后点击运算就可以看更新日志了""" % Version.VERSION)

UPDATE_LOG = """更新日志:
20220430-20220505
1.4.0
移出插件读取部分
干掉exec，解决命名空间污染
性能优化
ui调整
示例调整
修改了api和插件事件，文档更新
移除test模式

20220429
v1.3.3
debug

20220429
v1.3.2
添加了设置窗口
修复搜索框不可用的问题(完成移植工作)
修改保存位置的选项回归(完成移植工作)

宣告完成从wxpython移植到pyside6

20220416
v1.3.1
debug

20220414
v1.3.0
移植到pyside6上

20220412
v1.2.7
优化，更多的优化
主程序优化
插件优化

20220411
V1.2.6
增加了输出上限的选项
优化了输出优化的模式
修复了内置插件
修复遗留问题

20220409
V1.2.5
修复了遗留问题
完善了输出优化模式
细节修复
程序优化

20220408
V1.2.4
hpyculator模块上传了！现可通过pip下载
重写了内置插件
修改了gui,可自选保存位置

20220408
V1.2.3
修复了多线程的问题
修改了菜单栏

20220407
V1.2.2
修复了内置插件无法正常工作的问题
添加了更新检查按钮(打开浏览器跳转到https://github.com/HowieHz/hpyculator/releases)

20220407
V1.2.1
修复了无插件打不开的bug
修复了退出后还残留后台程序的bug

20220407
V1.2.0
修改gui
增加了搜索框
添加了测试选项，勾选之后程序就会使用test模式

20220405
V1.1.0
增加了对文件插件的支持
增加了很多提示，防止插件卡死程序

V1.0.0
改名为
皓式可编程计算器hpyculator
2022年1月19日添加了规范文档
2022年1月18-19日，重构了程序（从流水线+函数群变成了面向对象的编程）
2022年3月29再次捡起项目，
2022年3月31写完，实现了规范文档，但是掉进了打包的坑，还好第二天爬出来了
2022年4月1日第一个内测版
2022年4月1日添加了多线程支持，添加了更多用户友好类提示，修复了若干个bug，添加了一种输出模式
2022年4月2日修改了内置插件的性能，添加了新的输出模式，并提供了4个函数便于插件作者创作

趣闻:
2022年1月16日时想用PyQt（5/6）代替wxpython（哈哈wxpython不支持python3.10，其他版本又删掉了，被迫的）
由于PyQt（5/6）-tool用pip不能正确安装和考虑到学习成本，所以干脆卸载py310，改回py39（白配置一遍^_^）



来自 各类数组计算程序
2021年6月27日
V1.2
选择保存文件时不再输出
平均数，众数，中位数，方差，标准差的计算可以输入负数和小数了
为了节省体积，换了个图标

V1.1
添加了平均数，众数，中位数，方差，标准差的计算

2020年11月14日
V1.0
多了一个可以切换计算核心的选项
更加模块化，内置模块有杨辉三角计算和斐波那契数列计算
(作者电脑i5-5代低压u，4g内存。无聊算114514行，算了半小时，原本要关了，看任务管理器还在运行，就去睡了一会，起来看到一个1338412kb的文件（约1.28G），哈哈哈，请勿模仿（除非很闲）)



来自杨辉三角计算程序v3.4.2
a3.4.2 修复不能读取配置的问题，优化代码体积
2020年11月8日a3.4.1 创建了两个文件夹专门用来储存设置和输出,扩大1.5倍了默认生成的窗口
2020年11月8日a3.4.0 在输入栏输入update_log点击运算就可以看更新日志了!菜单栏它来了
2020年11月7日a3.3.4 修改保存名的时间
a3.3.3b 走向规范化
2020年11月7日a3.3.3a 遗弃的分支，消除了保存时间
2020年10月17日a3.3.2 删除多余代码优化体积
2020年10月17日a3.3.1 生成在gui里的不再是一行了,砍掉了可以修改算法的选项,删除多余代码优化体积
a3.3.0 被遗忘的版本号
a3.2 改进了算法，增快了运行速度,增加了可以修改算法的选项
2020年10月10日a3.1 将原来的运行并保存按钮修改为是否保存的选项
a3.0 有GUI了!生成的文件不再是一行了

a2.0 添加了保存文件到本目录，添加了计算计时器,防止因为错误信息造成的闪退,节省打包体积

a1.0 最初的经典"""
